src/duckling/canvas/tools/_grid-snap.ts

```diff
@@ -0,0 +1,88 @@
+import {Vector, vectorModulus, vectorAdd, vectorMultiply} from '../../math/vector';
+import {Box2} from '../../math/box2';
+
+/**
+ * All the distances corners would need to snap to for grid points
+ */
+interface CornersSnapDistance {
+    topLeft: Vector,
+    topRight: Vector,
+    bottomLeft: Vector,
+    bottomRight: Vector,
+    middle : Vector,
+    topMiddle : Vector,
+    rightMiddle : Vector,
+    bottomMiddle : Vector,
+    leftMiddle : Vector 
+};
+
+/**
+ * Get the minimum snap distance for the given corners
+ */
+export function minCornerSnapDistance(movementDestinationWithoutSnap : Vector, box : Box2, gridSize : Vector) : Vector {
+    let cornersSnapDistance = _cornersSnapDistance(movementDestinationWithoutSnap, box, gridSize);
```

Comment:
This code could be a lot simpler.

function minCornerSnapDistance(...) {
let snapDistance;
let minHypotenus = infinity;
for (let xDir of [-1, 0, 1]) {
    for (let yDir of [-1, 0, 1]) {
    let distance = _cornerSnapDistance({x: xDir, y: yDir}, desitnationWithoutSnap, box, gridSize);
    let hypotenus = _cornerSnapHypotenus(distance);
    if (hypotenus < minHypotenus) {
        minHypotenus = hypotenus;
        snapDistance = distance;
    }
}
return snapDistance;
}